File: java/triangle/src/main/java/Triangle.java
import java.util.stream.DoubleStream;

class Triangle {

    private final double[] sidesInAscendingOrder;

    Triangle(final double side1, final double side2, final double side3) throws TriangleException {
        sidesInAscendingOrder = DoubleStream.of(side1, side2, side3)
                .sorted()
                .toArray();

        if (!sidesHavePositiveLength()) {
            throw new TriangleException("all sides must be positive");
        }

        if (!hasTriangleInequality()) {
            throw new TriangleException("triangle inequality not fulfilled");
        }
    }

    boolean isEquilateral() {
        return sidesInAscendingOrder[0] == sidesInAscendingOrder[2];
    }

    boolean isIsosceles() {
        return sidesInAscendingOrder[0] == sidesInAscendingOrder[1] ||
                sidesInAscendingOrder[1] == sidesInAscendingOrder[2];
    }

    boolean isScalene() {
        return !isIsosceles();
    }

    private boolean hasTriangleInequality() {
        return sidesInAscendingOrder[2] <= (sidesInAscendingOrder[1] + sidesInAscendingOrder[0]);
    }

    private boolean sidesHavePositiveLength() {
        return sidesInAscendingOrder[0] > 0d;
    }
}

File: kotlin/triangle/src/main/kotlin/Triangle.kt
import kotlin.math.abs

class Triangle {
    private val numberOfDistinctSides: Int

    constructor(a: Int, b: Int, c: Int) : this(a.toDouble(), b.toDouble(), c.toDouble())

    constructor(a: Double, b: Double, c: Double) {
        require(a >= 0.0 && b >= 0.0 && c >= 0.0) { "a, b and c must not be 0" }
        require(abs(a - b) < c && c < a + b) { "triangle inequality not fulfilled" }

        numberOfDistinctSides = listOf(a, b, c).distinct().size
    }

    val isEquilateral: Boolean
        get() = numberOfDistinctSides == 1

    val isIsosceles: Boolean
        get() = numberOfDistinctSides < 3

    val isScalene: Boolean
        get() = numberOfDistinctSides == 3
}